### PR TITLE
Added required_features for issue #1570.

### PR DESCRIPTION
**Pull request by jbendig**
Based on PR #2056 by @tsurai and PR #2325 by @JanLikar 

I tried to fix most everything that was talked about in the previous pull requests. Docs still need to be updated though.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This looks fantastic and quite thorough to me, thanks @jbendig! I think with updates to the manifest docs this is good to go!

cc @rust-lang/tools, a neat feature being added to cargo!

**Comment by alexcrichton** [comment]
Oh and the Travis tests look to be failing, but you can just add `if !::is_nightly()` checks (or similar) to the tests

**Comment by jbendig** [comment]
Thanks @alexcrichton! I used your `!is_nightly()` advice to fix the tests. The nightly features are still so irresistible that I forgot about stable and running tests against it.

I added an explanation for *required-features* as a sub-section to **Configuring a target**. This seems like the best place because the toml block already there is more [lib] specific and **The [features] section** seems more about features in general. Any thoughts or preferences?

File: src/cargo/ops/cargo_compile.rs [hunk 1]
```diff
@@ -418,9 +425,37 @@ fn generate_targets<'a>(pkg: &'a Package,
...
+        }
+    };
+
+    //Collect the targets that are libraries or have all required features available.
+    let no_features = HashSet::new();
+    let features = features.unwrap_or(&no_features);
+    let mut compatible_targets = Vec::with_capacity(targets.len());
+    for (target, profile) in targets.drain(0..) {
+        if target.is_lib() || match target.required_features() {
+            Some(f) => !f.iter().any(|f| !features.contains(f)),
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Perhaps instead of the double-negation here we could go for `f.iter().all(...)`?

**Comment by alexcrichton** [comment]
Looks fantastic! 

Can you also add one more tests which emulates `cargo install` with `required-features`? Just to make sure that binaries not activated don't get installed and activated binaries do indeed get installed.

File: src/cargo/ops/cargo_compile.rs (hunk 1 above)
Comment (jbendig) [inline src/cargo/ops/cargo_compile.rs]: Sure, that's definitely more readable.

**Comment by jbendig** [comment]
Good idea. I'll try and make these changes later today.

**Comment by alexcrichton** [comment]
@bors: r+

Awesome, thanks @jbendig!

**Comment by pthariensflame** [comment]
Saw this on TWiR: how does this interact with `default` features?

**Comment by alexcrichton** [comment]
@pthariensflame it shouldn't interact directly, this reads activated features after they've been decided and then dictates whether the binaries/examples are compiled

**Comment by pthariensflame** [comment]
Oh, ok.

**Pull request merged**